src/sources/splunk_hec.rs

```diff
@@ -0,0 +1,888 @@
+    }
+
+    fn options() -> BoxedFilter<(Response<Body>,)> {
+        let post = warp::options()
+            .and(
+                warp::path::end()
+                    .or(path!("event").and(warp::path::end()))
+                    .or(path!("event" / "1.0").and(warp::path::end()))
+                    .or(path!("raw" / "1.0").and(warp::path::end()))
+                    .or(path!("raw").and(warp::path::end())),
+            )
+            .map(|_| warp::reply::with_header(warp::reply(), "Allow", "POST").into_response());
+
+        let get = warp::options()
+            .and(
+                (path!("health").and(warp::path::end()))
+                    .or(path!("health" / "1.0").and(warp::path::end())),
+            )
+            .map(|_| warp::reply::with_header(warp::reply(), "Allow", "GET").into_response());
+
+        post.or(get).unify().boxed()
+    }
+
+    /// Authorize request
+    fn authorization(&self) -> BoxedFilter<((),)> {
+        let credentials = self.credentials.clone();
+        warp::header::optional("Authorization")
+            .and_then(move |token: Option<String>| match token {
+                Some(token) if token.as_bytes() == credentials => Ok(()),
+                Some(_) => Err(Rejection::from(ApiError::InvalidAuthorization)),
+                None => Err(Rejection::from(ApiError::MissingAuthorization)),
+            })
+            .boxed()
+    }
+
+    /// Is body encoded with gzip
+    fn gzip(&self) -> BoxedFilter<(bool,)> {
+        warp::header::optional::<String>("Content-Encoding")
+            .and_then(|encoding: Option<String>| match encoding {
+                Some(s) if s.as_bytes() == b"gzip" => Ok(true),
+                Some(_) => Err(Rejection::from(ApiError::UnsupportedEncoding)),
+                None => Ok(false),
+            })
+            .boxed()
+    }
+
+    /// Sink shutdowns this source once source output is closed
+    fn sink_with_shutdown(&self) -> impl Sink<SinkItem = Event, SinkError = Rejection> + 'static {
+        let trigger = self.trigger.clone();
+        self.out.clone().sink_map_err(move |_| {
+            // Sink has been closed so server should stop listening
+            trigger
+                .try_lock()
+                .map(|mut lock| {
+                    // Stopping
+                    lock.take();
+                })
+                // If locking fails, that means someone else is stopping it.
+                .ok();
+
+            ApiError::ServerShutdown.into()
+        })
+    }
+}
+
+/// Constructs one ore more events from json-s coming from reader.
+/// If errors, it's done with input.
+struct EventStream<R: Read> {
+    /// Remaining request with JSON events
+    data: R,
+    /// Count of sended events
+    events: usize,
+    /// Optinal channel from headers
+    channel: Option<ValueKind>,
+    /// Extracted default time
+    time: Option<ValueKind>,
+    /// Remaining extracted default values
+    extractors: [DefaultExtractor; 4],
```

Comment:
I might be missing something but can you explain what these used for?

Reply:
Spluk HEC `api` specifies for some `json` fields that once they are seen their value becomes default for next events in the request until that field is found again. `DefaultExtractor` implements that behavior.